dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/TaskSetState.java

```diff
@@ -16,60 +16,135 @@
  */
 package com.google.edwmigration.dumper.application.dumper.task;
 
+import static com.google.edwmigration.dumper.application.dumper.task.TaskCategory.REQUIRED;
+import static com.google.edwmigration.dumper.application.dumper.task.TaskState.FAILED;
+
+import com.google.common.collect.ImmutableList;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.stream.Collectors;
 import javax.annotation.CheckForNull;
 import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
 import javax.annotation.concurrent.GuardedBy;
 import net.jcip.annotations.ThreadSafe;
 
 /** @author shevek */
 public interface TaskSetState {
 
+  @ParametersAreNonnullByDefault
+  static class TaskResultSummary {
+    private final String exception;
+    private final Task<?> task;
+    private final TaskState state;
+
+    private TaskResultSummary(String exception, Task<?> task, TaskState state) {
+      this.exception = exception;
+      this.task = task;
+      this.state = state;
+    }
+
+    @Nonnull
+    static TaskResultSummary create(Task<?> task, TaskResult<?> result) {
+      String optionalException = String.format(": %s", result.getException());
+      String exceptionDescription = result.getException() == null ? "" : optionalException;
+      return new TaskResultSummary(exceptionDescription, task, result.getState());
+    }
+
+    @Nonnull
+    @Override
+    public String toString() {
+      return String.format("Task %s (%s) %s%s", state, task.getCategory(), task, exception);
+    }
+  }
+
+  @ParametersAreNonnullByDefault
+  public static class TasksReport {
+
+    private final long count;
+    private final TaskState state;
+
+    @Nonnull
+    @Override
+    public String toString() {
+      return String.format("%d TASKS %s", count, state);
+    }
+
+    TasksReport(long count, TaskState state) {
+      this.count = count;
+      this.state = state;
+    }
+  }
+
   @ThreadSafe
   public static class Impl implements TaskSetState {
 
-    @GuardedBy("lock")
+    @GuardedBy("this")
     private final Map<Task<?>, TaskResult<?>> resultMap = new HashMap<>();
 
-    private final Object lock = new Object();
+    @Deprecated // Use TaskSetState instead of TaskSetState.Impl
+    @Override
+    public synchronized long failedRequiredTaskCount() {
+      long result = 0;
+      for (Task<?> key : resultMap.keySet()) {
```

Comment:
If both keys and values are needed, then I would suggest iterating over `entrySet`, not `keySet`, so that the reference to `resultMap` is not needed inside the loop.

Reply:
Done